File: android/src/main/java/com/plugin/pubnub/PubnubPlugin.java
package com.plugin.pubnub;

import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.EventChannel.EventSink;
import io.flutter.plugin.common.EventChannel.StreamHandler;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

import com.pubnub.api.PNConfiguration;
import com.pubnub.api.PubNub;

/** PubnubPlugin */
public class PubnubPlugin implements MethodCallHandler {
  /** Plugin registration. */
  public static void registerWith(Registrar registrar) {
    final MethodChannel channel = new MethodChannel(registrar.messenger(), "pubnub");
    channel.setMethodCallHandler(new PubnubPlugin());

    final EventChannel messageChannel = new EventChannel(registrar.messenger(), "plugins.flutter.io/pubnub_message");
    // final MessageStreamHandler messageStreamHandler = new MessageStreamHandler();
    messageChannel.setStreamHandler(new StreamHandler() {
      @Override
      public void onListen(Object arguments, EventSink events) {

      }

      @Override
      public void onCancel(Object arguments) {

      }
    });
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    switch (call.method) {
      case "getPlatformVersion":
        result.success("Android " + android.os.Build.VERSION.RELEASE);
      case "create":
        onCreate(call, result);
      default:
        result.notImplemented();
    }
  }

  public void onCreate(MethodCall call, Result result) {
    final String publishKey = call.argument("publishKey");
    final String subscribeKey = call.argument("subscribeKey");
    final String uuid = call.argument("uuid");

    if ((publishKey != null && !publishKey.isEmpty()) && (subscribeKey != null && !subscribeKey.isEmpty())) {
      PNConfiguration pnConfiguration = new PNConfiguration();
      pnConfiguration.setPublishKey(publishKey);
      pnConfiguration.setSubscribeKey(subscribeKey);
      pnConfiguration.setSecure(true);

      if (uuid != null && !uuid.isEmpty()) {
        pnConfiguration.setUuid(uuid);
      } else {
        pnConfiguration.setUuid(pnConfiguration.getUuid());
      }

      PubNub pubnub = new PubNub(pnConfiguration);
    }

    result.success(null);
  }
}

File: ios/Classes/PubnubPlugin.m
#import "PubnubPlugin.h"

@implementation PubnubPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"pubnub"
            binaryMessenger:[registrar messenger]];
  PubnubPlugin* instance = [[PubnubPlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
}

@end

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if ([@"getPlatformVersion" isEqualToString:call.method]) {
    result([@"iOS " stringByAppendingString:[[UIDevice currentDevice] systemVersion]]);
  } else {
    result(FlutterMethodNotImplemented);
  }
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    if  ([@"create" isEqualToString:call.method]) {
        result([self handleCreate:call]);
    } else {
        result(FlutterMethodNotImplemented);
    }
}

- (id) handleCreate:(FlutterMethodCall*)call {
    NSString *publishKey = call.arguments[@"publishKey"];
    NSString *subscribeKey = call.arguments[@"subscribeKey"];
    NSString *uuid = call.arguments[@"uuid"];
    if(publishKey && subscribeKey) {       
        self.config =
        [PNConfiguration configurationWithPublishKey:publishKey
                                        subscribeKey:subscribeKey];
        self.config.stripMobilePayload = NO;
        if(uuid) {
            self.config.uuid = uuid;
        } else {
            self.config.uuid = [NSUUID UUID].UUIDString.lowercaseString;
        }
        self.client = [PubNub clientWithConfiguration:self.config];
        [self.client addListener:self];
    }
    // Nothing to return, just return NULL
    return NULL;
}

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel
                                     methodChannelWithName:@"pubnub"
                                     binaryMessenger:[registrar messenger]];
    PubnubFlutterPlugin* instance = [[PubnubFlutterPlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
    // Event channel for streams
    instance.messageStreamHandler = [MessageStreamHandler new];
    FlutterEventChannel* messageChannel =
    [FlutterEventChannel eventChannelWithName:@"plugins.flutter.io/pubnub_message"
                              binaryMessenger:[registrar messenger]];
    [messageChannel setStreamHandler:instance.messageStreamHandler];
    // Event channel for streams
    instance.statusStreamHandler = [StatusStreamHandler new];
    FlutterEventChannel* statusChannel =
    [FlutterEventChannel eventChannelWithName:@"plugins.flutter.io/pubnub_status"
                              binaryMessenger:[registrar messenger]];
    [statusChannel setStreamHandler:instance.statusStreamHandler];
    instance.errorStreamHandler = [ErrorStreamHandler new];
}

@implementation MessageStreamHandler
- (FlutterError*)onListenWithArguments:(id)arguments eventSink:(FlutterEventSink)eventSink {
    self.eventSink = eventSink;
    return nil;
}
- (FlutterError*)onCancelWithArguments:(id)arguments {
    self.eventSink = nil;
    return nil;
}
- (void) sendMessage:(PNMessageResult *)message {
     if(self.eventSink) {
         NSDictionary *result = @{@"uuid": message.uuid, @"channel": message.data.channel, @"message": message.data.message};
         self.eventSink(result);
     }
}

@end

@implementation StatusStreamHandler
- (FlutterError*)onListenWithArguments:(id)arguments eventSink:(FlutterEventSink)eventSink {
    self.eventSink = eventSink;
    return nil;
}
- (FlutterError*)onCancelWithArguments:(id)arguments {
    self.eventSink = nil;
    return nil;
}
- (void) sendStatus:(PNStatus *)status {
    if(self.eventSink) {
        self.eventSink(status.stringifiedOperation);
    }
}

@end